### PR TITLE
Experiment utils

### PR DESCRIPTION
**Pull request by EmmaRenauld**
- LossHistoryMonitor: Biggest change. I don't think we need to keep track of the count: can be found with len(...). 

Other changes:
- Compute_gradient_norm. NOTE. THERE WAS A MATH ERROR. See if it changes your work so far. Corrected.
- Early stopping monitor: simply added more description
- memory_log: ok
- Timer: simply added more description.
- IterTimer: removed.

### CODE REVIEW

File: dwi_ml/experiment/monitoring.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
+# -*- coding: utf-8 -*-
 import numpy as np
 
 
+# Checked!
 class LossHistoryMonitor(object):
```
Comment (ppoulin91) [inline dwi_ml/experiment/monitoring.py]: I set it up in my code as a `ValueMonitor`, since we can use it to monitor any kind of metric while training

File: dwi_ml/experiment/monitoring.py [hunk 2]
```diff
@@ -18,62 +20,75 @@ class LossHistoryMonitor(object):
...
     def __init__(self, name):
         self.name = name
-        self.history = []
-        self.epochs = []
-        self.sum = 0.0
-        self.count = 0
-        self._avg = 0.0
-        self.num_iter = 0
-        self.num_epochs = 0
+        self.history = []  # The list of loss values
```
Comment (ppoulin91) [inline dwi_ml/experiment/monitoring.py]: What do you think of renaming `history` as `updates`, and `current_history` as `current_epoch_updates`? I think it would help for clarity.

File: dwi_ml/experiment/monitoring.py [hunk 3]
```diff
@@ -18,62 +20,75 @@ class LossHistoryMonitor(object):
...
+        self.epochs = []  # The list of averaged loss values per epoch
+        self.current_history = [] # The list of loss values in current epoch
+
+    @property
+    def total_iter_count(self):
```
Comment (ppoulin91) [inline dwi_ml/experiment/monitoring.py]: Regarding last comment, this could become `num_updates`

File: dwi_ml/experiment/monitoring.py [hunk 4]
```diff
@@ -18,62 +20,75 @@ class LossHistoryMonitor(object):
...
+        self.epochs.append(np.mean(self.current_history))
+        self.current_history = []
 
     def get_state(self):
         return {'name': self.name,
                 'history': self.history,
-                'epochs': self.epochs,
-                'num_iter': self.num_iter,
-                'num_epochs': self.num_epochs}
+                'epochs': self.epochs}
```
Comment (ppoulin91) [inline dwi_ml/experiment/monitoring.py]: We should keep track of `current_history` also

File: dwi_ml/experiment/monitoring.py [hunk 5]
```diff
@@ -18,62 +20,75 @@ class LossHistoryMonitor(object):
...
 
     def set_state(self, state):
         self.name = state['name']
```
Comment (ppoulin91) [inline dwi_ml/experiment/monitoring.py]: Reminder to set `current_history` if we save it in `get_state`

File: dwi_ml/experiment/learning_utils.py [hunk 6]
```diff
@@ -16,5 +21,5 @@ def compute_gradient_norm(parameters):
     for p in parameters:
         param_norm = p.grad.as_tensor.norm(2)
         total_norm += param_norm.item() ** 2
-    total_norm = total_norm ** (1. / 3)
+    total_norm = total_norm ** (1. / 2)
```
Comment (ppoulin91) [inline dwi_ml/experiment/learning_utils.py]: Good catch! Luckily, this function was only used for monitoring, and did not affect any computation

File: dwi_ml/experiment/monitoring.py (hunk 1 above)
Comment (EmmaRenauld) [reply dwi_ml/experiment/monitoring.py]: ok

File: dwi_ml/experiment/learning_utils.py (hunk 6 above)
Comment (EmmaRenauld) [reply dwi_ml/experiment/learning_utils.py]: ok, good!
